### PR TITLE
Data assess doc

### PR DESCRIPTION
**Pull request by amrasekh**
- Added dataset assessment documentations 
- Updated DatasetModule parameters design for simplicity and flexibility
- Updated sensitive feature predictability ROC-AUC score to use for categorical features too
- Created dataset assessment demo to walk a user through the assessment steps, algorithms, and results interpretations

### CODE REVIEW

File: credoai/modules/dataset_modules/dataset_base.py [hunk 1]
```diff
@@ -68,91 +95,135 @@ def _group_differences(self):
         return diffs
     
     def _run_cv(self, pipe):
+        """Determines cross-validated ROC-AUC score
```
Comment (IanAtCredo) [inline credoai/modules/dataset_modules/dataset_base.py:98]: This docstring should mention "redundant encoding". That's really what this is testing for

https://towardsdatascience.com/programming-fairness-in-algorithms-4943a13dd9f8

File: docs/notebooks/README.md [hunk 2]
```diff
@@ -12,11 +12,14 @@ Lens is primarily a framework for comprehensive assessment of AI models. However
 
 The [Connection](https://github.com/credo-ai/credoai_lens/blob/develop/docs/notebooks/governance_integration.ipynb) notebook walks you through the steps to set up the connection. One connected, there are a lot that can be communicated between the two. In the follow-up [Integration](https://github.com/credo-ai/credoai_lens/blob/develop/docs/notebooks/integration_demo.ipynb) demo, we will take a model created and assessed completely independently of Lens and send that data to Credo AI's Governance Platform.
```
Comment (IanAtCredo) [inline docs/notebooks/README.md:13]: Make sure to update `tutorials.rst` as well!

**Review by IanAtCredo — CHANGES_REQUESTED** [review]
Left some comments. Haven't yet run the whole thing, which I will do later today - so I can't yet give feedback on the demo notebook itself. There are a few conflicts with the just-merged visualization PR as well.

Highest level thing is I'm not sure what we gained from changing the module to take in data instead of X, y and sensitive features. It doesn't _really_ matter, so feel free to keep it as it is, but I thought we spoke about not changing it, so this was a surprise. I'm not sure what we gained.

**Comment by IanAtCredo** [comment]
Went through demo notebook, which was awesome. Loved the section breaking down the different functions.

It broke on the predicting the sensitive feature with xgboost. I fixed that error. I also wanted to get feature importance from that model, which was difficult because of the preprocessing (column transformer). I found some code that lets us get the feature names out of the preprocessor and used that to add "sensitive_feature_prediction_feature_importances" to the results of the dataset assessment. This helps a lot!

File: credoai/utils/dataset_utils.py [hunk 3]
```diff
@@ -23,4 +28,50 @@ def concat_features_label_to_dataframe(X, y, sensitive_features):
         label_name = 'label'
         df[label_name] = sensitive_features
 
-    return df, sensitive_feature_name, label_name
\ No newline at end of file
+    return df, sensitive_feature_name, label_name
+
+class ColumnTransformerUtil:
```
Comment (IanAtCredo) [inline credoai/utils/dataset_utils.py:33]: @amrasekh , here is the new code that gets feature names from the column transformer

File: credoai/modules/dataset_modules/dataset_base.py [hunk 4]
```diff
@@ -58,101 +70,170 @@ def prepare_results(self):
...
+                             error_score='raise')
+        # get feature importances by running once
+        pipe.fit(X, sensitive_features)
+        model = pipe['model']
+        preprocessor = pipe['preprocessor']
+        col_names = ColumnTransformerUtil.get_ct_feature_names(preprocessor)
+        feature_importances = pd.Series(model.feature_importances_, 
+            index=col_names).sort_values(ascending=False)
+        results['sensitive_feature_prediction_score'] = cv_results.mean()
+        results['sensitive_feature_prediction_feature_importances'] = feature_importances
```
Comment (IanAtCredo) [inline credoai/modules/dataset_modules/dataset_base.py:131]: Added feature iimoprtances

File: credoai/modules/dataset_modules/dataset_base.py [hunk 5]
```diff
@@ -58,101 +70,170 @@ def prepare_results(self):
...
+
+        Returns
+        -------
+        ndarray
+            Cross-validation score
+        """
+        results = {}
+        pipe = self._make_pipe()
+        X = self.data.drop(columns=[self.sensitive_feature_key, self.label_key])
+        sensitive_features = self.data[self.sensitive_feature_key].cat.codes
```
Comment (IanAtCredo) [inline credoai/modules/dataset_modules/dataset_base.py:115]: Using `cat.codes` which gets out numerical representations of the sensitive feature. Necessary for xgboost.

File: credoai/lens.py [hunk 6]
```diff
@@ -348,36 +347,87 @@ class CredoData:
...
+        self.label_key = label_key
+        self.categorical_features_keys = categorical_features_keys
+
+        self.name = name
+        self.sensitive_features = data[sensitive_feature_key]
+        self.y = data[label_key]
+        self.X = data.drop(columns=[label_key], axis=1)
+
+        if omit_for_model_features_keys:
+            self.X = self.X.drop(columns=omit_for_model_features_keys, axis=1, errors='ignore')
```
[the remark this answers is not in the document]
Comment (amrasekh) [reply credoai/lens.py]: Renamed it to `unused_features_keys`
I don't think so. If a user has used the sensitive feature for training, then it wouldn't be unused, which means they would leave `unused_features_keys` just as is, i.e. `None`.

File: credoai/lens.py [hunk 7]
```diff
@@ -348,36 +347,87 @@ class CredoData:
...
+    data : pd.DataFrame
+        Dataset dataframe that includes all features and labels
+    sensitive_feature_key : str
+        Name of the sensitive feature column, like 'race' or 'gender'
+    label_key : str
+        Name of the label column
+    categorical_features_keys : list[str], optional
+        Names of the categorical features
+    omit_for_model_features_keys : list[str], optional
+        Names of the features to not use for model training
```
[the remark this answers is not in the document]
Comment (amrasekh) [reply credoai/lens.py]: Updated as suggested

File: credoai/modules/dataset_modules/dataset_base.py [hunk 8]
```diff
@@ -1,52 +1,65 @@
...
+    label_key : str
+        Name of the label column
+    categorical_features_keys : list[str], optional
+        Names of the categorical features (including the sensitive feature, if applicable)
     """    
     def __init__(self,
-                 X,
-                 y,
-                 sensitive_features):                
-        self.X = pd.DataFrame(X)
```
[the remark this answers is not in the document]
Comment (amrasekh) [reply credoai/modules/dataset_modules/dataset_base.py:36]: There was a lot of dataframe aggregations was going on, all of which has been removed with this switch. Also a function was needed for them, that is no longer needed and was removed.
Agreed that creating `X` many times was not ideal. It is now created one in `init`. 
The class as a whole is more readable and has less redundancy and verbosity now.

File: credoai/modules/dataset_modules/dataset_base.py [hunk 9]
```diff
@@ -68,91 +95,135 @@ def _group_differences(self):
...
-        float_cols = df.select_dtypes(include=[np.float]).columns
-        cat_cols = set(cols) - set(float_cols)
-
-        for name, column in df.iteritems():
-            unique_count = column.unique().shape[0]
-            total_count = column.shape[0]
-            if unique_count / total_count < threshold:
-                cat_cols.add(name)
+        """
+        X = self.data.drop(columns=[self.label_key])
```
[the remark this answers is not in the document]
Comment (amrasekh) [reply credoai/modules/dataset_modules/dataset_base.py]: Please see the response above.

File: credoai/modules/dataset_modules/dataset_base.py [hunk 10]
```diff
@@ -68,91 +95,135 @@ def _group_differences(self):
...
+        Returns
+        -------
+        sklearn.pipeline
+            Pipeline of scaler and model transforms
+        """
+        # Estimate categorical features, if not provided
+        if self.categorical_features_keys:
+            categorical_features = self.categorical_features_keys.copy()
+        else:
+            categorical_features = self._find_categorical_features()
```
[the remark this answers is not in the document]
Comment (amrasekh) [reply credoai/modules/dataset_modules/dataset_base.py]: Updated as suggested

File: credoai/reporting/dataset_base.py [hunk 11]
```diff
@@ -53,9 +53,7 @@ def _plot_balance_metrics(self):
         # Generate sample balance barplots
         results = results_all["balance_metrics"]["sample_balance"]
         df = pd.DataFrame(results)
-        sensitive_feature_name = list(df.drop(["count", "percentage"], axis=1).columns)[
-            0
-        ]
+        sensitive_feature_name = results_all["meta_data"]["sensitive_feature_key"]
```
[the remark this answers is not in the document]
Comment (amrasekh) [reply credoai/reporting/dataset_base.py]: It is now rolled back. meta_data has also been removed from results.

File: credoai/modules/dataset_modules/dataset_base.py [hunk 12]
```diff
@@ -1,52 +1,65 @@
...
 
 import pandas as pd
 
 class DatasetModule(CredoModule):
     """Dataset module for Credo AI. 
-
-    This module takes in features and labels and provides functionality to:
-    - perform proxy analysis of features
-    - generate fairness-oriented descriptive information 
+    This module takes in features and labels and provides functionality to perform dataset assessment
```
[the remark this answers is not in the document]
Comment (amrasekh) [reply credoai/modules/dataset_modules/dataset_base.py:20]: Change applied everywhere.

File: credoai/modules/dataset_modules/dataset_base.py (hunk 1 above)
Comment (amrasekh) [reply credoai/modules/dataset_modules/dataset_base.py]: docstring updated

File: docs/notebooks/README.md (hunk 2 above)
Comment (amrasekh) [reply docs/notebooks/README.md:13]: Update made

File: credoai/lens.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (IanAtCredo) [reply credoai/lens.py]: The primary use case of this assessment is to evaluate the fairness characteristics of the dataset. Almost all of the time the team _will_ remove the explicit sensitive feature. I feel this is optimizing for an edge case rather than the majority use. The end result is that almost every time the user will have to list that the sensitive feature should not be included in X. At the very least I'd like a default argument that adds the sensitive feature to `unused_feature_keys`. We can talk during standup tomorrow.

**Review by IanAtCredo — APPROVED** [review]
I believe we should default to removing sensitive features quite strongly. Besides that change, everything else looks great!

**Comment by IanAtCredo** [comment]
Updated. 

As I was going through it again I still don't think the module should take in `data`. It lead to unexpected performance.

For instance, when I include the sensitive feature in CredoData's X, I should see that it is one of the features in the dataset module's mutual information plot. And the ROC of the overall proxy detection should be 1. Instead, those features are dropped from the dataset, without my knowledge!

The module is making choices, presuming that I removed the sensitive feature and label from my data. This prevents the module from catching some real issues! If my X data truly does contain a sensitive feature, I should see that... I have a proxy for the sensitive feature in my X.

I believe the module could be changed to accept X, y and sensitive features with minimal changes to the code, while being more predictable. Right now it's not making use of the preprocessing we do in CredoData, which is breaking some modularity. We want all modules to interact with the same data.
